### PR TITLE
Profile the node as part of the initial connection

### PR DESCRIPTION
**Pull request by WilliamMcCumstie**
When the daemon starts in now uses `lshw` and `/proc/meminfo` to profile the node for cpu, gpu, and memory information. This is sent as part of the `CONNECTED` message.

Whilst re-connections will resend the information, the profiler will only run once. The daemon needs to be restarted for the profiler to run again.

### CODE REVIEW

File: spec/batch_script_runner_spec.rb [hunk 1]
```diff
@@ -45,30 +45,30 @@
     )
   end
 
-  it { should be_valid }
+  xit { should be_valid }
```
Comment (benarmston) [inline spec/batch_script_runner_spec.rb]: There are lots of specs that have been marked as pending.  Do these no longer work?  If they don't is that due to work on this branch?

File: lib/flight_scheduler/errors.rb [hunk 2]
```diff
@@ -28,4 +28,5 @@
 module FlightScheduler
   class BaseError < RuntimeError; end
   class JobValidationError < BaseError; end
+  class ProfilerError < RuntimeError; end
```
Comment (benarmston) [inline lib/flight_scheduler/errors.rb]: For errors that are only raised from a single class, we should standardise on namespacing them in the class that raises them.  Inheriting from a common base error is fine.

File: lib/flight_scheduler/application.rb [hunk 3]
```diff
@@ -56,21 +56,40 @@ def connection
...
+
+    # NOTE: This method can not use the MessageSender as it could block
+    #       the main task from re-establishing the connection.
+    def send_connected(con)
+      con.write({
+        command: 'CONNECTED',
+        auth_token: auth_token,
+        cpus: profiler.cpus,
+        gpus: profiler.gpus,
+        memory: profiler.memory
```
Comment (benarmston) [inline lib/flight_scheduler/application.rb:75]: Can we get some logging of the profile result.  Either here or in the profiler class.  Having that logging will make debugging issues much easier.

File: spec/fixtures/profiler/Standard_NC24s_v3.metadata.yaml [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+memory: 473392132096
+
+# The Azure documentation says there are four GPUs for this type of machine
+# However on inspection of lshw output, there are actually five:
+#  * four NVIDIA Corporation - GV100GL [Tesla V100 PCIe 16GB], and
+#  * one Microsoft Corporation - VGA compatible controller
+#
+# At the moment all five GPUs are consider equals, which in reality is not
+# the case. Consider revisiting at a future point
+gpus: 5
```
Comment (benarmston) [inline spec/fixtures/profiler/Standard_NC24s_v3.metadata.yaml:11]: Agreed. This is good enough for now.

**Review by benarmston — CHANGES_REQUESTED** [review]
This looks pretty good to me.  I like your use of test fixtures here :-)

Worthy of note is that correct profiling requires `lshw` to be ran as root.  As the daemon also needs to be able to execute processes as arbitrary users, there is already a requirement for the daemon to be ran as root.  So this is OK.

I've added a few comments that should be quick to address.

File: spec/batch_script_runner_spec.rb (hunk 1 above)
Comment (WilliamMcCumstie) [reply spec/batch_script_runner_spec.rb]: All these specs where broken before this was implemented

File: spec/fixtures/profiler/Standard_NC24s_v3.metadata.yaml (hunk 4 above)
Comment (WilliamMcCumstie) [reply spec/fixtures/profiler/Standard_NC24s_v3.metadata.yaml:11]: :+1:

File: spec/batch_script_runner_spec.rb (hunk 1 above)
Comment (benarmston) [reply spec/batch_script_runner_spec.rb]: Fair enough. I'll fix them and push a commit to this branch.  'Cos I just can't merge a branch with so many pending specs.  Apparently merging branches with broken specs is fine though /whistles/.

Comment (benarmston) [reply spec/batch_script_runner_spec.rb]: I've added a commit to get those tests working @WilliamMcCumstie .

**Comment by benarmston** [comment]
Thanks for those changes @WilliamMcCumstie.  This looks good to me now.

**Pull request merged**